wgpu-core/src/command/query.rs

```diff
@@ -0,0 +1,441 @@
+    Use(#[from] QueryUseError),
+    #[error("Error encountered while trying to resolve a query")]
+    Resolve(#[from] ResolveError),
+    #[error("Buffer {0:?} is invalid or destroyed")]
+    InvalidBuffer(id::BufferId),
+    #[error("QuerySet {0:?} is invalid or destroyed")]
+    InvalidQuerySet(id::QuerySetId),
+}
+
+/// Error encountered while trying to use queries
+#[derive(Clone, Debug, Error)]
+pub enum QueryUseError {
+    #[error("Query {query_index} is out of bounds for a query set of size {query_set_size}")]
+    OutOfBounds {
+        query_index: u32,
+        query_set_size: u32,
+    },
+    #[error("Query {query_index} has already been used within the same renderpass. Queries must only be used once per renderpass")]
+    UsedTwiceInsideRenderpass { query_index: u32 },
+    #[error("Query {new_query_index} was started while query {active_query_index} was already active. No more than one statistic or occlusion query may be active at once")]
+    AlreadyStarted {
+        active_query_index: u32,
+        new_query_index: u32,
+    },
+    #[error("Query was stopped while there was no active query")]
+    AlreadyStopped,
+    #[error("A query of type {query_type:?} was started using a query set of type {set_type:?}")]
+    IncompatibleType {
+        set_type: SimplifiedQueryType,
+        query_type: SimplifiedQueryType,
+    },
+}
+
+/// Error encountered while trying to resolve a query.
+#[derive(Clone, Debug, Error)]
+pub enum ResolveError {
+    #[error("Queries can only be resolved to buffers that contain the COPY_DST usage")]
+    MissingBufferUsage,
+    #[error("Resolving queries {start_query}..{end_query} would overrun the query set of size {query_set_size}")]
+    QueryOverrun {
+        start_query: u32,
+        end_query: u32,
+        query_set_size: u32,
+    },
+    #[error("Resolving queries {start_query}..{end_query} ({stride} byte queries) will end up overruning the bounds of the destination buffer of size {buffer_size} using offsets {buffer_start_offset}..{buffer_end_offset}")]
+    BufferOverrun {
+        start_query: u32,
+        end_query: u32,
+        stride: u32,
+        buffer_size: BufferAddress,
+        buffer_start_offset: BufferAddress,
+        buffer_end_offset: BufferAddress,
+    },
+}
+
+fn validate_query<'a, B: GfxBackend, E>(
+    tracker: &mut ResourceTracker<PhantomData<id::QuerySetId>>,
+    storage: &'a Storage<QuerySet<B>, id::QuerySetId>,
+    query_type: SimplifiedQueryType,
+    query_set_id: id::QuerySetId,
+    query_index: u32,
+    reset_state: &mut Option<&mut QueryResetMap>,
+    invalid_query_set_fn: impl FnOnce(id::QuerySetId) -> E,
+) -> Result<(&'a QuerySet<B>, hal::query::Query<'a, B>), E>
+where
+    E: From<QueryUseError>,
+{
+    let query_set = tracker
+        .use_extend(storage, query_set_id, (), ())
+        .map_err(|e| match e {
+            UseExtendError::InvalidResource => invalid_query_set_fn(query_set_id),
```

Comment:
I find the need to pass `invalid_query_set_fn` around to be unfortunate, and I'm thinking if we can get around it.
What if, instead of having `tracker`, `storage`, `query_set_id`, and `invalid_query_set_fn` parameters on a free function like this, we would do:
```rust
impl QuerySet {
  fn validate_query(&self, query_type, query_index, reset_state) -> Result<hal::query::Query<'a, B>,  QueryUseError> {...}
  fn write_timestamp(&self, ...) ...
  fn begin_pipeline_statistics_query(&self, ...) ...
  fn end_pipeline_statistics_query() ...
}
```

We'd then use the `tracker.use_extend` call in all the spots where this function is called, explicitly.

Benefits are:
  - less argument complexity on the functions, especially the 
  - easier to import `use crate::resource::QuerySet` instead of `use crate::resource::{fun_a, fun_b, fun_c};`
  - the calls to `use_extend` are more visible and consistent in the code (arguable, I guess?)